src/midi/soundcanvas.cpp

```diff
@@ -280,10 +284,60 @@ static void setup_filter(MixerChannelPtr& channel, const bool filter_enabled)
 	}
 }
 
+// Check whether a file with a specified extension exists at a specific
+// sub directory depth
+static bool filetype_at_depth(std_fs::path const& root_dir, int depth, std::string const& extension)
+{
+	std::error_code ec;
+	if (!std_fs::is_directory(root_dir, ec)) {
+		return false;
+	}
+	for (auto dir_iter = std_fs::recursive_directory_iterator(root_dir, ec);
+			dir_iter != std_fs::recursive_directory_iterator(); ++dir_iter) {
+		if (ec) {
+			return false;
+		}
+		if (dir_iter.depth() == depth) {
+			auto entry = *dir_iter;
+			if (entry.is_regular_file() && entry.path().extension() == extension) {
+				return true;
+			}
+		}
+	}
+	return false;
+}
+
 MidiDeviceSoundCanvas::MidiDeviceSoundCanvas()
 {
 	using namespace SoundCanvas;
 
+	// Before attempting to load any plugins, we must set an environment
+	// variable for the ROM directory if possible. Note, on POSIX systems 
+	// getenv/setenv are NOT thread safe
+	std::error_code ec;
+	bool rom_dir_found = false;
+	for (auto rom_dir : get_rom_dirs(get_soundcanvas_section()->GetString("soundcanvas_rom_dir"),
+									 DefaultSoundcanvasRomsDir,
+									 sc_platform_rom_dir_name,
+									 sc_platform_rom_dir_name_macos)) {
+		// Each sound canvas romset is in its own subdirectory.
+		// Attempt to find at least one .bin file in a subdirectory of 
+		// the current rom dir.
+		if (filetype_at_depth(rom_dir, 1, ".bin")) {
+			const auto canonical_rom_dir = std_fs::canonical(rom_dir, ec);
+			if (ec) {
+				break;
+			}
+			LOG_MSG("SOUNRCANVAS: Setting SOUNDCANVAS_ROM_DIR env variable to '%s'", canonical_rom_dir.string().c_str());
+			set_env_var("SOUNDCANVAS_ROM_DIR", canonical_rom_dir.string().c_str());
+			rom_dir_found = true;
+			break;
+		}
+	}
+	if (!rom_dir_found) {
+		LOG_MSG("SOUNRCANVAS: %s", "ROM dir not found, using plugin defaults");
+	}
```

Comment:
Sorry again, let's not do this.

We don't want to "leak knowledge" about the CLAP plugin's internals into the Staging codebase. Just set the env var, init the CLAP plugin, and report and error in the Staging logs if initialising the plugin failed. That's all we need to do here!

We don't wanna detect anything, we're just giving a hint to the CLAP plugin via the env var. We don't wanna encode knowledge here how the CLAP plugin stuctures its data files, what the extensions of those files are, and so on. The CLAP plugin should be a black box to Staging.

Reply:
Are you suggesting I get rid of the `filetype_at_depth()` stuff? I'm fine with that, I was of two minds about keeping that anyway.

Reply:
Yes please, that can go 👍🏻

Reply:
Actually, just trying to think how this would work. I still need to know which directory to fall back to.

Maybe I'll just choose the first non-empty directory.

Reply:
That's for the best, yeah.